IBX-414 - Measurement FT description (#1626)
ezsystems/developer-documentation: docs/api/field_types_reference/measurementfield.md

# Measurement Field Type

The Measurement Field Type represents measurement information. 
It stores the unit of measure, and either a single measurement value, 
or a pair of values that defines a range.

| Name          | Internal name       | Expected input type                                |
|---------------|---------------------|----------------------------------------------------|
| `Measurement` | `ibexa_measurement` | `Ibexa\Contracts\Measurement\Value\ValueInterface` |

## PHP API Field Type

### Input expectations

> To create a value, a Service of `Ibexa\Contracts\Measurement\MeasurementServiceInterface` 
> type is required.

Reviewer: ```suggestion
To create a value, a service implementing `Ibexa\Contracts\Measurement\MeasurementServiceInterface` 
is required.
```
not sure about this, but I don't think we talk about "being a type of interface"?

> You must inject the service directly with [dependency injection(../api/service_container.md). 

Author: ```suggestion
You must inject the service directly with [dependency injection](../api/service_container.md). 
```

|`\Ibexa\Contracts\Measurement\Value\SimpleValueInterface`| `$measurementService->buildSimpleValue('length', 2.5, 'centimeter')`|
|`\Ibexa\Contracts\Measurement\Value\RangeValueInterface` | `$measurementService->buildRangeValue('length', 1.2, 4.5,  'inch')` |

### Value object

##### Properties

The Value class of this Field Type contains the following properties:

| Property | Type                                             | Description                                                                                                          |
|----------|--------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `$value` |`Ibexa\Contracts\Measurement\Value\ValueInterface`| Stores the Measurement API Value, which can be either an instance of `Ibexa\Contracts\Measurement\Value\SimpleValueInterface` or `Ibexa\Contracts\Measurement\Value\RangeValueInterface`. |

##### Constructor

> The `Measurement\Value`constructor for this Value object initializes a new Value 

Reviewer: ```suggestion
The `Measurement\Value` constructor for this value object initializes a new value 
```

object with the value provided. 
As its first argument it accepts an object of `Ibexa\Contracts\Measurement\Value\ValueInterface` type.

Depending on the selected input type, the object resembles the following examples:

``` php
// Simple input (single value) example

// @var MeasurementServiceInterface $measurementService

// Instantiates a Measurement Value object
$measurementValue = new Measurement\Value(
    $measurementService->buildSimpleValue(
                    'length',
                    13.5,
                    'centimeter'
    )
);
```

``` php
// Range input value example

// @var MeasurementServiceInterface $measurementService

// Instantiates a Measurement Value object
$measurementValue = new Measurement\Value(
    $measurementService->buildRangeValue(
        'volume',
        0.5,
        0.7,
        'liter'
    )
);
```

### Validation

The Measurement Field Type validates measurement types and units passed within 

> the Value object against a list of the ones that the system supports, which can 

Reviewer: ```suggestion
the value object against a list of the ones that the system supports, which can 
```

be found in the `vendor/ibexa/measurement/src/bundle/Resources/config/builtin_units.yaml` file.

> ### Extending the default list of Measurement types and units

Reviewer: How about:
```suggestion
### Add new Measurement types and units
```

Author: It is not only about adding new units, it is also about overriding existing unit names.

> ### Extending the default list of Measurement types and units

Reviewer: I think this would work best as a separate topic, maybe somewhere alongside https://doc.ibexa.co/en/master/extending/add_user_setting/

Author: Yes, I considered placing this information in a separate article but Andrew insists we keep it here, as it will most likely be invalidated by features that he is currently working on.

> You can extend the default list of Measurement types and units by modifying the existing entries or adding new ones. To do this, you modify your YAML configuration, for example, by creating the `config/packages/ibexa_measurement.yaml` file.

Reviewer: ```suggestion
You can extend the default list of Measurement types and units by modifying the existing entries or adding new ones. 
To do this, modify your YAML configuration, for example, by creating a `config/packages/ibexa_measurement.yaml` file.
```

To override an existing designation of the unit of measure by changing the symbol that corresponds to a nautical unit of speed, and to add a rotational speed unit, add the following lines to your YAML configuration:

```yaml
ibexa_measurement:
    types:
        speed:
            knot: { symbol: kt }
            revolutions per minute: { symbol: RPM }

ibexa:
    system:
        default:
            measurement:
                types:
                    speed:
                        - revolutions per minute
```

To add a new Measurement type with its own new units, add the following lines to your YAML configuration:

```yaml
ibexa_measurement:
    types:
        my_type:
            my_unit: { symbol: my }
ibexa:
    system:
        default:
            measurement:
                types:
                    my_type:
                        - my_unit
```

!!! note

>     To be available for selection in the Back Office, each new Measurement type or unit must be enabled for the SiteAccess, in which it is intended to be used.

Reviewer: ```suggestion
    To be available for selection in the Back Office, each new Measurement type or unit must be enabled for the  Back Office SiteAccess.
```
